internal/executor/service/job_utilisation_reporter.go

```diff
@@ -0,0 +1,146 @@
+	dataAccessMutex   sync.Mutex
+	reportingInterval time.Duration
+}
+
+type podUtilisationInfo struct {
+	lastReported   time.Time
+	pod            *v1.Pod
+	utilisationMax common.ComputeResources
+}
+
+func NewUtilisationEventReporter(
+	clusterContext clusterContext.ClusterContext,
+	podUtilisation PodUtilisationService,
+	eventReporter reporter.EventReporter,
+	reportingPeriod time.Duration,
+) *UtilisationEventReporter {
+
+	r := &UtilisationEventReporter{
+		clusterContext:    clusterContext,
+		podUtilisation:    podUtilisation,
+		eventReporter:     eventReporter,
+		reportingInterval: reportingPeriod,
+		podInfo:           map[string]*podUtilisationInfo{},
+	}
+
+	clusterContext.AddPodEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod, ok := obj.(*v1.Pod)
+			if !ok {
+				log.Errorf("Failed to process pod event due to it being an unexpected type. Failed to process %+v", obj)
+				return
+			}
+			go r.updatePod(pod)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newPod, ok := newObj.(*v1.Pod)
+			if !ok {
+				log.Errorf("Failed to process pod event due to it being an unexpected type. Failed to process %+v", newObj)
+				return
+			}
+			go r.updatePod(newPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod, ok := obj.(*v1.Pod)
+			if !ok {
+				log.Errorf("Failed to process pod event due to it being an unexpected type. Failed to process %+v", obj)
+				return
+			}
+			go r.deletePod(pod)
+		},
+	})
+	return r
+}
+
+func (r *UtilisationEventReporter) ReportUtilisationEvents() {
+	r.dataAccessMutex.Lock()
+	defer r.dataAccessMutex.Unlock()
+
+	now := time.Now()
+	reportingTime := now.Add(-r.reportingInterval)
+	for _, info := range r.podInfo {
+		currentUtilisation := r.podUtilisation.GetPodUtilisation(info.pod)
+		info.utilisationMax.Max(currentUtilisation)
+		if info.lastReported.Before(reportingTime) {
+			r.reportUsage(info)
+			info.lastReported = now
+		}
+	}
+}
+
+func (r *UtilisationEventReporter) updatePod(pod *v1.Pod) {
+	if !util.IsManagedPod(pod) {
+		return
+	}
+
+	r.dataAccessMutex.Lock()
+	defer r.dataAccessMutex.Unlock()
+
+	if pod.Status.Phase == v1.PodRunning {
+		_, exists := r.podInfo[pod.Name]
+		if !exists {
+			r.podInfo[pod.Name] = &podUtilisationInfo{
+				lastReported:   time.Now(),
+				pod:            pod,
+				utilisationMax: r.podUtilisation.GetPodUtilisation(pod),
+			}
+		}
+	}
+	if util.IsInTerminalState(pod) {
+		podInfo, exists := r.podInfo[pod.Name]
+		if exists {
+			r.reportUsage(podInfo)
+			delete(r.podInfo, pod.Name)
+		}
+	}
+}
+
+func (r *UtilisationEventReporter) deletePod(pod *v1.Pod) {
+	if !util.IsManagedPod(pod) {
+		return
+	}
+	delete(r.podInfo, pod.Name)
+}
+
+func (r *UtilisationEventReporter) reportUsage(info *podUtilisationInfo) {
+	event := reporter.CreateJobUtilisationEvent(info.pod, info.utilisationMax, r.clusterContext.GetClusterId())
+	r.queueEventWithRetry(event, 3)
+	info.utilisationMax = common.ComputeResources{}
```

Comment:
Should resetting info fields be done together. Currently we set reported time in ReportUtilisationEvents and utilisationMax here.

Would be nice if they could be together, so state isn't modified as much in various places.